packages/tlschema/src/records/TLShape.ts

```diff
@@ -200,16 +188,136 @@ export function getShapePropKeysByStyle(props: Record<string, T.Validatable<any>
 	return propKeysByStyle
 }
 
+export const NO_DOWN_MIGRATION = 'none' as const
+// If a down migration was deployed more than a couple of months ago it should be safe to retire it.
+// We only really need them to smooth over the transition between versions, and some folks do keep
+// browser tabs open for months without refreshing, but at a certain point that kind of behavior is
+// on them. Plus anyway recently chrome has started to actually kill tabs that are open for too long rather
+// than just suspending them, so if other browsers follow suit maybe it's less of a concern.
+export const RETIRED_DOWN_MIGRATION = 'retired' as const
+
+/**
+ * @public
+ */
+export type TLShapePropsMigrations = {
```

Comment:
in my record-ful bindings spike, i ended up pulling a lot of this stuff out into a file for generic "record with props" migrations.

having sat with it a while though, i actually think the thing i would maybe do here is maybe remove these shape-specific migration helpers & just have everything be record migration sequences... or maybe even normal sequences with record-level migrations being the default, and an optional default `recordFilter` or something so you don't have to filter everything.

one thing that would give us is a way to make the store migration to split bindings out of arrows as part of the sequence for arrows themselves, which avoids some of the awkward stuff around depending on a sequence that doesn't exist.

It's maybe pushing a little additional complexity onto users for that "writing migrations for a custom shape" use case, but also potentially simplifies the story a fair bit and unlocks some useful options. What do you think?

Reply:
I've made it so that folks can use a normal migration sequence on the shapeUtil.migrations property but I'm not sure I like the idea of removing the props helpers stuff because it will indeed add a lot of complexity for the most common use case: folks who just wanna migrate their shape props will have to know about filters and namespaces and scopes, and doing everything inside the .props object manually.

I'm going to go with the current thing for now. There's still time to rethink before our next sdk release.